### PR TITLE
Fix data race in query finish/cancel

### LINKED ISSUE #42346: ThreadSanitizer: data race between query finish and cancel
```
WARNING: ThreadSanitizer: data race (pid=661)
  Write of size 8 at 0x7b6800590188 by thread T350:
    #0 memset <null> (clickhouse+0xb89f44d) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #1 std::__1::vector<std::__1::shared_ptr<DB::IProcessor>, std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>>::vector(std::__1::vector<std::__1::shared_ptr<DB::IProcessor>, std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>>&&) build_docker/../contrib/libcxx/include/vector:1194:31 (clickhouse+0x1ac8fb6f) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #2 DB::QueryPipeline::QueryPipeline(DB::QueryPipeline&&) build_docker/../src/QueryPipeline/QueryPipeline.cpp:33:16 (clickhouse+0x1ac8fb6f)
    #3 DB::QueryPipeline::reset() build_docker/../src/QueryPipeline/QueryPipeline.cpp:537:31 (clickhouse+0x1ac8fb6f)
    #4 DB::BlockIO::onFinish() build_docker/../src/QueryPipeline/BlockIO.h:40:18 (clickhouse+0x1d22f087) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #5 DB::TCPHandler::runImpl() build_docker/../src/Server/TCPHandler.cpp:403:26 (clickhouse+0x1d22f087)
    #6 DB::TCPHandler::run() build_docker/../src/Server/TCPHandler.cpp:1884:9 (clickhouse+0x1d2420c7) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #7 Poco::Net::TCPServerConnection::start() build_docker/../contrib/poco/Net/src/TCPServerConnection.cpp:43:3 (clickhouse+0x20a8fd62) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #8 Poco::Net::TCPServerDispatcher::run() build_docker/../contrib/poco/Net/src/TCPServerDispatcher.cpp:115:20 (clickhouse+0x20a905d2) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #9 Poco::PooledThread::run() build_docker/../contrib/poco/Foundation/src/ThreadPool.cpp:199:14 (clickhouse+0x20ce8239) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #10 Poco::(anonymous namespace)::RunnableHolder::run() build_docker/../contrib/poco/Foundation/src/Thread.cpp:55:11 (clickhouse+0x20ce664f) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #11 Poco::ThreadImpl::runnableEntry(void*) build_docker/../contrib/poco/Foundation/src/Thread_POSIX.cpp:345:27 (clickhouse+0x20ce4c87) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)

  Previous read of size 8 at 0x7b6800590188 by thread T359 (mutexes: write M0, write M1):
    #0 std::__1::vector<std::__1::shared_ptr<DB::IProcessor>, std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>>::begin() build_docker/../contrib/libcxx/include/vector:1408:30 (clickhouse+0x1d2af3d7) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #1 DB::ExecutingGraph::cancel() build_docker/../src/Processors/Executors/ExecutingGraph.cpp:389:27 (clickhouse+0x1d2af3d7)
    #2 DB::PipelineExecutor::cancel() build_docker/../src/Processors/Executors/PipelineExecutor.cpp:73:12 (clickhouse+0x1d2a6aa0) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #3 DB::QueryStatus::cancelQuery(bool) build_docker/../src/Interpreters/ProcessList.cpp:382:12 (clickhouse+0x1bed782f) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #4 DB::ProcessList::sendCancelToQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, bool) build_docker/../src/Interpreters/ProcessList.cpp:486:18 (clickhouse+0x1bed782f)
    #5 DB::SyncKillQuerySource::generate() build_docker/../src/Interpreters/InterpreterKillQueryQuery.cpp:164:42 (clickhouse+0x1bdc15d4) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #6 DB::ISource::tryGenerate() build_docker/../src/Processors/ISource.cpp:124:18 (clickhouse+0x1d290d54) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #7 DB::ISource::work() build_docker/../src/Processors/ISource.cpp:94:26 (clickhouse+0x1d290789) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #8 DB::executeJob(DB::ExecutingGraph::Node*, DB::ReadProgressCallback*) build_docker/../src/Processors/Executors/ExecutionThreadContext.cpp:47:26 (clickhouse+0x1d2b65e8) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #9 DB::ExecutionThreadContext::executeTask() build_docker/../src/Processors/Executors/ExecutionThreadContext.cpp:92:9 (clickhouse+0x1d2b65e8)
    #10 DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) build_docker/../src/Processors/Executors/PipelineExecutor.cpp:228:26 (clickhouse+0x1d2a7aa0) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #11 DB::PipelineExecutor::executeSingleThread(unsigned long) build_docker/../src/Processors/Executors/PipelineExecutor.cpp:194:5 (clickhouse+0x1d2a6dc2) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #12 DB::PipelineExecutor::executeImpl(unsigned long) build_docker/../src/Processors/Executors/PipelineExecutor.cpp:367:9 (clickhouse+0x1d2a6dc2)
    #13 DB::PipelineExecutor::execute(unsigned long) build_docker/../src/Processors/Executors/PipelineExecutor.cpp:89:9 (clickhouse+0x1d2a6b99) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #14 DB::threadFunction(DB::PullingAsyncPipelineExecutor::Data&, std::__1::shared_ptr<DB::ThreadGroupStatus>, unsigned long) build_docker/../src/Processors/Executors/PullingAsyncPipelineExecutor.cpp:79:24 (clickhouse+0x1d2babd4) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #15 DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0::operator()() const build_docker/../src/Processors/Executors/PullingAsyncPipelineExecutor.cpp:107:13 (clickhouse+0x1d2babd4)
    #16 decltype(static_cast<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&>(fp)()) std::__1::__invoke_constexpr<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&) build_docker/../contrib/libcxx/include/type_traits:3648:23 (clickhouse+0x1d2babd4)
    #17 decltype(auto) std::__1::__apply_tuple_impl<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&, std::__1::tuple<>&>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&, std::__1::tuple<>&, std::__1::__tuple_indices<>) build_docker/../contrib/libcxx/include/tuple:1595:1 (clickhouse+0x1d2babd4)
    #18 decltype(auto) std::__1::apply<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&, std::__1::tuple<>&>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&, std::__1::tuple<>&) build_docker/../contrib/libcxx/include/tuple:1604:1 (clickhouse+0x1d2babd4)
    #19 ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'()::operator()() build_docker/../src/Common/ThreadPool.h:193:13 (clickhouse+0x1d2babd4)
    #20 decltype(static_cast<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(fp)()) std::__1::__invoke<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'()&>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&) build_docker/../contrib/libcxx/include/type_traits:3640:23 (clickhouse+0x1d2babd4)
    #21 void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'()&>(ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'()&) build_docker/../contrib/libcxx/include/__functional/invoke.h:61:9 (clickhouse+0x1d2babd4)
    #22 std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'(), void ()>::operator()() build_docker/../contrib/libcxx/include/__functional/function.h:230:12 (clickhouse+0x1d2babd4)
    #23 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/../contrib/libcxx/include/__functional/function.h:711:16 (clickhouse+0x1d2babd4)
    #24 std::__1::__function::__policy_func<void ()>::operator()() const build_docker/../contrib/libcxx/include/__functional/function.h:843:16 (clickhouse+0x1326d54e) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #25 std::__1::function<void ()>::operator()() const build_docker/../contrib/libcxx/include/__functional/function.h:1184:12 (clickhouse+0x1326d54e)
    #26 ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) build_docker/../src/Common/ThreadPool.cpp:294:17 (clickhouse+0x1326d54e)
    #27 void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/../src/Common/ThreadPool.cpp:144:73 (clickhouse+0x13271b51) (BuildId: 34dcdceaa0c18fb90c4378c415f7da4e0b1622f0)
    #28 decltype(static_cast<void>(fp)()) std::__1::__invoke<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) build_docker/../contrib/libcxx/include/type_traits:3640:23 (clickhouse+0x13271b51)
    #29 void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>) build_docker/../contrib/libcxx/include/thread:282:5 (clickhouse+0x13271b51)
    #30 void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, long, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) build_docker/../contrib/libcxx/include/thread:293:5 (clickhouse+0x13271b51)
```

https://s3.amazonaws.com/clickhouse-test-reports/42190/352f57d1058def63fd35e3e7f9aa7c318d9d711d/stateless_tests__tsan__[1/3].html

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix a data race in query finish/cancel. This closes #42346.

See https://s3.amazonaws.com/clickhouse-test-reports/42325/980f79f9a72aa994f26fac8a6a1e3d0791b5a182/stateless_tests__tsan__s3_storage__[1/3]/stderr.log

### CODE REVIEW

File: src/QueryPipeline/QueryPipeline.cpp [hunk 1]
```diff
@@ -529,7 +530,7 @@ void QueryPipeline::addCompletedPipeline(QueryPipeline other)
         throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot add not completed pipeline");
 
     resources = std::move(other.resources);
-    processors.insert(processors.end(), other.processors.begin(), other.processors.end());
+    processors->insert(processors->end(), other.processors->begin(), other.processors->end());
 }
 
 void QueryPipeline::reset()
```
Comment (alexey-milovidov) [inline src/QueryPipeline/QueryPipeline.cpp:536]: Here it will simply reset the shared_ptr instead of modifying the shared std::vector.

**Comment by alexey-milovidov** [comment]
The test successfully reproduced the data race on the previous version.

**Comment by alexey-milovidov** [comment]
Unfortunately, another race condition has been uncovered by the new test.

**Comment by alexey-milovidov** [comment]
@azat please help with reviewing.

**Review by azat — COMMENTED** [review]
LGTM.

But one more thing is possible right now, here is an [ASan report](https://s3.amazonaws.com/clickhouse-test-reports/42362/e93b8d1618da04678fc3c54afb2646516be81d8d/stateless_tests__asan__[2/2]/stderr.log)

<details>

<summary>ASan report</summary>

```
==642==ERROR: AddressSanitizer: heap-use-after-free on address 0x61c0000176f0 at pc 0x0000420772b9 bp 0x7fdad7ab9c00 sp 0x7fdad7ab9bf8
WRITE of size 4 at 0x61c0000176f0 thread T3 (TCPHandler)
    #0 0x420772b8 in std::__1::__shared_mutex_base::unlock() build_docker/../contrib/libcxx/src/shared_mutex.cpp:54:14
    #1 0x420772b8 in std::__1::shared_timed_mutex::unlock() build_docker/../contrib/libcxx/src/shared_mutex.cpp:111:53
    #2 0x31397d71 in std::__1::unique_lock<std::__1::shared_timed_mutex>::~unique_lock() build_docker/../contrib/libcxx/include/__mutex_base:141:19
    #3 0x31397d71 in DB::StorageFileSink::~StorageFileSink() build_docker/../src/Storages/StorageFile.cpp:741:7
    #4 0x2d422cac in std::__1::__shared_count::__release_shared() build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:174:9
    #5 0x2d422cac in std::__1::__shared_weak_count::__release_shared() build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:216:27
    #6 0x2d422cac in std::__1::shared_ptr<DB::IProcessor>::~shared_ptr() build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:703:23
    #7 0x2d422cac in void std::__1::__destroy_at<std::__1::shared_ptr<DB::IProcessor>, 0>(std::__1::shared_ptr<DB::IProcessor>*) build_docker/../contrib/libcxx/include/__memory/construct_at.h:56:13
    #8 0x2d422cac in void std::__1::destroy_at<std::__1::shared_ptr<DB::IProcessor>, 0>(std::__1::shared_ptr<DB::IProcessor>*) build_docker/../contrib/libcxx/include/__memory/construct_at.h:81:5
    #9 0x2d422cac in void std::__1::allocator_traits<std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>>::destroy<std::__1::shared_ptr<DB::IProcessor>, void, void>(std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>&, std::__1::shared_ptr<DB::IProcessor>*) build_docker/../contrib/libcxx/include/__memory/allocator_traits.h:317:9
    #10 0x2d422cac in std::__1::vector<std::__1::shared_ptr<DB::IProcessor>, std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>>::__base_destruct_at_end(std::__1::shared_ptr<DB::IProcessor>*) build_docker/../contrib/libcxx/include/vector:803:9
    #11 0x2d422cac in std::__1::vector<std::__1::shared_ptr<DB::IProcessor>, std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>>::__clear() build_docker/../contrib/libcxx/include/vector:797:29
    #12 0x2d422cac in std::__1::vector<std::__1::shared_ptr<DB::IProcessor>, std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>>::~vector() build_docker/../contrib/libcxx/include/vector:401:9
    #13 0x32a8738a in std::__1::__shared_count::__release_shared() build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:174:9
    #14 0x32a8738a in std::__1::__shared_weak_count::__release_shared() build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:216:27
    #15 0x32a8738a in std::__1::shared_ptr<std::__1::vector<std::__1::shared_ptr<DB::IProcessor>, std::__1::allocator<std::__1::shared_ptr<DB::IProcessor>>>>::~shared_ptr() build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:703:23
    #16 0x32a8738a in DB::ExecutingGraph::~ExecutingGraph() build_docker/../src/Processors/Executors/ExecutingGraph.h:14:7
    #17 0x32a7d273 in std::__1::default_delete<DB::ExecutingGraph>::operator()(DB::ExecutingGraph*) const build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:54:5
    #18 0x32a7d273 in std::__1::unique_ptr<DB::ExecutingGraph, std::__1::default_delete<DB::ExecutingGraph>>::reset(DB::ExecutingGraph*) build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:315:7
    #19 0x32a7d273 in std::__1::unique_ptr<DB::ExecutingGraph, std::__1::default_delete<DB::ExecutingGraph>>::~unique_ptr() build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:269:19
    #20 0x32a7d273 in DB::PipelineExecutor::~PipelineExecutor() build_docker/../src/Processors/Executors/PipelineExecutor.cpp:62:1
    #21 0x32a7c376 in std::__1::__shared_count::__release_shared() build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:174:9
    #22 0x32a7c376 in std::__1::__shared_weak_count::__release_shared() build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:216:27
    #23 0x32a7c376 in std::__1::shared_ptr<DB::PipelineExecutor>::~shared_ptr() build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:703:23
    #24 0x32a7c376 in DB::CompletedPipelineExecutor::Data::~Data() build_docker/../src/Processors/Executors/CompletedPipelineExecutor.cpp:31:5
    #25 0x32a7b3a9 in std::__1::default_delete<DB::CompletedPipelineExecutor::Data>::operator()(DB::CompletedPipelineExecutor::Data*) const build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:54:5
    #26 0x32a7b3a9 in std::__1::unique_ptr<DB::CompletedPipelineExecutor::Data, std::__1::default_delete<DB::CompletedPipelineExecutor::Data>>::reset(DB::CompletedPipelineExecutor::Data*) build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:315:7
    #27 0x32a7b3a9 in std::__1::unique_ptr<DB::CompletedPipelineExecutor::Data, std::__1::default_delete<DB::CompletedPipelineExecutor::Data>>::~unique_ptr() build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:269:19
    #28 0x32a7b3a9 in DB::CompletedPipelineExecutor::~CompletedPipelineExecutor() build_docker/../src/Processors/Executors/CompletedPipelineExecutor.cpp:115:1
    #29 0x32981d05 in DB::TCPHandler::runImpl() build_docker/../src/Server/TCPHandler.cpp:411:13
    #30 0x329b1c62 in DB::TCPHandler::run() build_docker/../src/Server/TCPHandler.cpp:1884:9
    #31 0x3a2a106e in Poco::Net::TCPServerConnection::start() build_docker/../contrib/poco/Net/src/TCPServerConnection.cpp:43:3
    #32 0x3a2a1df2 in Poco::Net::TCPServerDispatcher::run() build_docker/../contrib/poco/Net/src/TCPServerDispatcher.cpp:115:20
    #33 0x3a7772db in Poco::PooledThread::run() build_docker/../contrib/poco/Foundation/src/ThreadPool.cpp:199:14
    #34 0x3a770bc7 in Poco::ThreadImpl::runnableEntry(void*) build_docker/../contrib/poco/Foundation/src/Thread_POSIX.cpp:345:27
    #35 0x7fdbe4dad608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x8608) (BuildId: 7b4536f41cdaa5888408e82d0836e33dcf436466)
    #36 0x7fdbe4cd2132 in __clone (/lib/x86_64-linux-gnu/libc.so.6+0x11f132) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)

0x61c0000176f0 is located 1648 bytes inside of 1728-byte region [0x61c000017080,0x61c000017740)
freed by thread T3 (TCPHandler) here:
    #0 0xe005f12 in operator delete(void*, unsigned long) (/usr/bin/clickhouse+0xe005f12) (BuildId: fa20781d8151c5e133eeffa9373a343950a7245b)
    #1 0x2cf20ff4 in std::__1::__shared_weak_count::__release_shared() build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:217:9
    #2 0x2cf20ff4 in std::__1::shared_ptr<DB::IStorage>::~shared_ptr() build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:703:23
    #3 0x2cf20ff4 in void std::__1::__destroy_at<std::__1::shared_ptr<DB::IStorage>, 0>(std::__1::shared_ptr<DB::IStorage>*) build_docker/../contrib/libcxx/include/__memory/construct_at.h:56:13
    #4 0x2cf20ff4 in void std::__1::destroy_at<std::__1::shared_ptr<DB::IStorage>, 0>(std::__1::shared_ptr<DB::IStorage>*) build_docker/../contrib/libcxx/include/__memory/construct_at.h:81:5
    #5 0x2cf20ff4 in void std::__1::allocator_traits<std::__1::allocator<std::__1::shared_ptr<DB::IStorage>>>::destroy<std::__1::shared_ptr<DB::IStorage>, void, void>(std::__1::allocator<std::__1::shared_ptr<DB::IStorage>>&, std::__1::shared_ptr<DB::IStorage>*) build_docker/../contrib/libcxx/include/__memory/allocator_traits.h:317:9
    #6 0x2cf20ff4 in std::__1::vector<std::__1::shared_ptr<DB::IStorage>, std::__1::allocator<std::__1::shared_ptr<DB::IStorage>>>::__base_destruct_at_end(std::__1::shared_ptr<DB::IStorage>*) build_docker/../contrib/libcxx/include/vector:803:9
    #7 0x2cf20ff4 in std::__1::vector<std::__1::shared_ptr<DB::IStorage>, std::__1::allocator<std::__1::shared_ptr<DB::IStorage>>>::__clear() build_docker/../contrib/libcxx/include/vector:797:29
    #8 0x2cf20ff4 in std::__1::vector<std::__1::shared_ptr<DB::IStorage>, std::__1::allocator<std::__1::shared_ptr<DB::IStorage>>>::~vector() build_docker/../contrib/libcxx/include/vector:401:9
    #9 0x2d434b63 in DB::QueryPlanResourceHolder::~QueryPlanResourceHolder() build_docker/../src/QueryPipeline/PipelineResourcesHolder.cpp:21:51
    #10 0x2d448fef in DB::QueryPipeline::reset() build_docker/../src/QueryPipeline/QueryPipeline.cpp:546:1
    #11 0x32981cc9 in DB::BlockIO::onFinish() build_docker/../src/QueryPipeline/BlockIO.h:39:18
    #12 0x32981cc9 in DB::TCPHandler::runImpl() build_docker/../src/Server/TCPHandler.cpp:403:26
    #13 0x329b1c62 in DB::TCPHandler::run() build_docker/../src/Server/TCPHandler.cpp:1884:9
    #14 0x3a2a106e in Poco::Net::TCPServerConnection::start() build_docker/../contrib/poco/Net/src/TCPServerConnection.cpp:43:3
    #15 0x3a2a1df2 in Poco::Net::TCPServerDispatcher::run() build_docker/../contrib/poco/Net/src/TCPServerDispatcher.cpp:115:20
    #16 0x3a7772db in Poco::PooledThread::run() build_docker/../contrib/poco/Foundation/src/ThreadPool.cpp:199:14
    #17 0x3a770bc7 in Poco::ThreadImpl::runnableEntry(void*) build_docker/../contrib/poco/Foundation/src/Thread_POSIX.cpp:345:27
    #18 0x7fdbe4dad608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x8608) (BuildId: 7b4536f41cdaa5888408e82d0836e33dcf436466)
```

</details>

Maybe move `CompletedPipelineExecutor` into a separate scope? https://github.com/ClickHouse/ClickHouse/blob/ddfaadeea4577bdd378a74ad1ed43e06345383e4/src/Server/TCPHandler.cpp#L393

**Comment by azat** [comment]
>[Stateless tests flaky check (asan) — fail: 7, passed: 1](https://s3.amazonaws.com/clickhouse-test-reports/42362/ee5f5a4cb488c9f3be38c03684ea0256b4e3815c/stateless_tests__release_.html)

CI found one more issue:

```
2022.10.21 10:19:16.235137 [ 646 ] {} <Fatal> BaseDaemon: Code: 49. DB::Exception: Context has expired. (LOGICAL_ERROR), Stack trace (when copying this message, always include the lines below):
2022.10.21 10:19:16.235148 [ 646 ] {} <Fatal> BaseDaemon: 
2022.10.21 10:19:16.235165 [ 646 ] {} <Fatal> BaseDaemon: 0. ./build_docker/../src/Common/Exception.cpp:69: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xcd9451a in /usr/lib/debug/usr/bin/clickhouse.debug
2022.10.21 10:19:16.235185 [ 646 ] {} <Fatal> BaseDaemon: 1. DB::WithContextImpl<std::__1::shared_ptr<DB::Context const> >::getContext() const @ 0x7f9b608 in /usr/lib/debug/usr/bin/clickhouse.debug
2022.10.21 10:19:16.235198 [ 646 ] {} <Fatal> BaseDaemon: 2. ./build_docker/../src/Interpreters/ProcessList.cpp:0: DB::QueryStatus::~QueryStatus() @ 0x1254848a in /usr/lib/debug/usr/bin/clickhouse.debug
2022.10.21 10:19:16.235220 [ 646 ] {} <Fatal> BaseDaemon: 3. ./build_docker/../contrib/libcxx/src/include/atomic_support.h:74: DB::ProcessList::sendCancelToQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool) @ 0x12549207 in /usr/lib/debug/usr/bin/clickhouse.debug
2022.10.21 10:19:16.235235 [ 646 ] {} <Fatal> BaseDaemon: 4. ./build_docker/../src/Interpreters/InterpreterKillQueryQuery.cpp:166: DB::SyncKillQuerySource::generate() @ 0x12463ced in /usr/lib/debug/usr/bin/clickhouse.debug
2022.10.21 10:19:16.235248 [ 646 ] {} <Fatal> BaseDaemon: 5. ./build_docker/../src/Processors/Chunk.h:90: DB::ISource::tryGenerate() @ 0x133d9215 in /usr/lib/debug/usr/bin/clickhouse.debug
2022.10.21 10:19:16.235265 [ 646 ] {} <Fatal> BaseDaemon: 6. ./build_docker/../contrib/libcxx/include/optional:321: DB::ISource::work() @ 0x133d8da6 in /usr/lib/debug/usr/bin/clickhouse.debug
2022.10.21 10:19:16.235278 [ 646 ] {} <Fatal> BaseDaemon: 7. ./build_docker/../src/Processors/Executors/ExecutionThreadContext.cpp:0: DB::ExecutionThreadContext::executeTask() @ 0x133f4ee6 in /usr/lib/debug/usr/bin/clickhouse.debug
2022.10.21 10:19:16.235291 [ 646 ] {} <Fatal> BaseDaemon: 8. ./build_docker/../src/Processors/Executors/PipelineExecutor.cpp:229: DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) @ 0x133e905c in /usr/lib/debug/usr/bin/clickhouse.debug
2022.10.21 10:19:16.235304 [ 646 ] {} <Fatal> BaseDaemon: 9. ./build_docker/../contrib/libcxx/include/__memory/shared_ptr.h:702: DB::PipelineExecutor::executeImpl(unsigned long) @ 0x133e7a79 in /usr/lib/debug/usr/bin/clickhouse.debug
2022.10.21 10:19:16.235317 [ 646 ] {} <Fatal> BaseDaemon: 10. ./build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:284: DB::PipelineExecutor::execute(unsigned long) @ 0x133e787d in /usr/lib/debug/usr/bin/clickhouse.debug
2022.10.21 10:19:16.235332 [ 646 ] {} <Fatal> BaseDaemon: 11. ./build_docker/../src/Processors/Executors/PullingAsyncPipelineExecutor.cpp:0: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0>(DB::PullingAsyncPipelineExecutor::pull(DB::Chunk&, unsigned long)::$_0&&)::'lambda'(), void ()> >(std::__1::__function::__policy_storage const*) @ 0x133f8849 in /usr/lib/debug/usr/bin/clickhouse.debug
2022.10.21 10:19:16.235347 [ 646 ] {} <Fatal> BaseDaemon: 12. ./build_docker/../base/base/wide_integer_impl.h:772: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xce4f18c in /usr/lib/debug/usr/bin/clickhouse.debug
2022.10.21 10:19:16.235360 [ 646 ] {} <Fatal> BaseDaemon: 13. ./build_docker/../contrib/libcxx/include/__memory/unique_ptr.h:312: void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>, bool)::'lambda0'()> >(void*) @ 0xce5489e in /usr/lib/debug/usr/bin/clickhouse.debug
```

Now it is possible for QueryStatus to destroyed after `ProcessListEntry::~ProcessListEntry`

**Comment by KochetovNicolai** [comment]
Integration tests (tsan) -  #42575, not connected

**Pull request merged**